jax/core.py

```diff
@@ -751,7 +751,7 @@ class AbstractValue:
   __slots__: List[str] = []
 
   def at_least_vspace(self):
-    assert False
+    return self
```

Comment:
I'm calling `at_least_vspace` in a few places now (like `JVPTrace`) to instantiate tangent avals of the correct dtype, but this results in sometimes hitting the `assert False` in `AbstractValue` here. As a quick fix this now does `return self`. 
I'm not sure what the `assert False` is guarding against: maybe it doesn't make sense to reason about the vspace of an `AbstractValue` because it shouldn't have a tangent value? Any objection to removing the assert?

Reply:
I think it was meant to be something like a "raise NotImplementedError", or like an abstract method on an abstract base class. Using `return self` could be a sensible default, though I'm interested in what errors pop up if we leave the `assert False`, just to investigate whether there are other solutions...

Reply:
For example in `api_test` this happens in ``

```
```

Reply:
For example in `api_test` the assertion is hit in CustomJVPTest.test_initial_style, CustomVJPTest.test_initial_style and RematTest.test_remat_symbolic_zeros.  In these tests it hits a `full_raise` on a `JVPTrace`. I think before this change the `full_raise` will return a `JVPTrace` with a `Zero(AbstractValue)` as a tangent.

I've left in the assert if you want to test it out.

End of the CustomJVPTest.test_initial_style stacktrace:

```
 File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/interpreters/partial_eval.py", line 139, in process_primitive
    return custom_partial_eval_rules[primitive](self, *tracers, **params)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/lax/lax_control_flow.py", line 1593, in _scan_partial_eval
    unroll=unroll)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/lax/lax_control_flow.py", line 1829, in scan_bind
    return core.Primitive.bind(scan_p, *args, **params)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 276, in bind
    tracers = map(top_trace.full_raise, args)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/util.py", line 35, in safe_map
    return list(map(f, *args))
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 372, in full_raise
    return self.pure(val)
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/interpreters/ad.py", line 243, in pure
    tangent_zero = Zero(get_aval(val).at_least_vspace())
  File "/build/work/cc0082b7d2d0e2638dfb9604ed2f62baf8cc/google3/runfiles/google3/third_party/py/jax/core.py", line 754, in at_least_vspace
    assert False
AssertionError
```